File: backend/reconstruct.js
const fs = require('fs');
const readline = require('readline');
const path = require('path');

async function replay() {
  const rl = readline.createInterface({
    input: fs.createReadStream('C:\\Users\\dines\\.gemini\\antigravity-ide\\brain\\bf135cc7-99ee-4e1d-a2d3-a51e74cf3f05\\.system_generated\\logs\\transcript.jsonl')
  });

  const files = {};
  for await (const line of rl) {
    const step = JSON.parse(line);
    if (step.step_index >= 2390) break;

    if (step.tool_calls) {
      for (const call of step.tool_calls) {
        if (['write_to_file', 'replace_file_content', 'multi_replace_file_content'].includes(call.name)) {
            const rawArgs = call.args;
            const targetRaw = rawArgs.TargetFile || rawArgs.targetFile;
            if (!targetRaw) continue;
            let target = targetRaw;
            if (targetRaw.startsWith('"')) target = JSON.parse(targetRaw);

            target = target.replace(/\\\\/g, '\\').replace(/\\\\\\\\/g, '\\').replace(/\\\\/g, '\\');
            
            if (target.endsWith('Dashboard.js') || target.endsWith('app.py') || target.endsWith('OrderDetail.js')) {
                if (!files[target]) files[target] = fs.readFileSync(target, 'utf8');

                if (call.name === 'write_to_file') {
                    let code = rawArgs.CodeContent || rawArgs.codeContent;
                    if (code.startsWith('"')) code = JSON.parse(code);
                    files[target] = code;
                } else if (call.name === 'replace_file_content') {
                    let code = rawArgs.ReplacementContent || rawArgs.replacementContent;
                    if (code.startsWith('"')) code = JSON.parse(code);
                    const lines = files[target].split('\n');
                    let start = rawArgs.StartLine;
                    let end = rawArgs.EndLine;
                    if (typeof start === 'string' && start.startsWith('"')) start = JSON.parse(start);
                    if (typeof end === 'string' && end.startsWith('"')) end = JSON.parse(end);
                    start = parseInt(start) - 1;
                    end = parseInt(end);
                    lines.splice(start, end - start, ...(code.split('\n')));
                    files[target] = lines.join('\n');
                } else if (call.name === 'multi_replace_file_content') {
                    let chunks = rawArgs.ReplacementChunks || rawArgs.replacementChunks;
                    if (typeof chunks === 'string') chunks = JSON.parse(chunks);
                    chunks.sort((a, b) => parseInt(b.StartLine || b.startLine) - parseInt(a.StartLine || a.startLine));
                    let lines = files[target].split('\n');
                    for (const chunk of chunks) {
                        let code = chunk.ReplacementContent || chunk.replacementContent;
                        let start = chunk.StartLine || chunk.startLine;
                        let end = chunk.EndLine || chunk.endLine;
                        const s = parseInt(start) - 1;
                        const e = parseInt(end);
                        lines.splice(s, e - s, ...(code.split('\n')));
                    }
                    files[target] = lines.join('\n');
                }
            }
        }
      }
    }
  }

  for (const [target, content] of Object.entries(files)) {
      console.log('Writing reconstructed', target);
      fs.writeFileSync(target, content, 'utf8');
  }
}
replay().then(() => console.log('Done')).catch(console.error);

File: backend/reconstruct.py
import json
import os

transcript_path = r"C:\Users\dines\.gemini\antigravity-ide\brain\bf135cc7-99ee-4e1d-a2d3-a51e74cf3f05\.system_generated\logs\transcript.jsonl"
target_step = 2390

files = {}

def get_base_file(target):
    if target not in files:
        try:
            with open(target, 'r', encoding='utf-8') as f:
                files[target] = f.read()
        except FileNotFoundError:
            files[target] = ""

with open(transcript_path, 'r', encoding='utf-8') as f:
    for line in f:
        try:
            step = json.loads(line)
        except:
            continue
            
        step_idx = step.get('step_index', 0)
        if step_idx >= target_step:
            break
            
        calls = step.get('tool_calls', [])
        for call in calls:
            name = call.get('name')
            if name in ('write_to_file', 'replace_file_content', 'multi_replace_file_content'):
                args = call.get('args', {})
                if isinstance(args, str):
                    try: args = json.loads(args)
                    except: continue
                        
                target = args.get('TargetFile') or args.get('targetFile')
                if not target:
                    continue
                    
                while isinstance(target, str) and target.startswith('"'):
                    try: target = json.loads(target)
                    except: break
                    
                target = target.replace('\\\\', '\\')
                
                if target.endswith('Dashboard.js') or target.endswith('app.py') or target.endswith('OrderDetail.js') or target.endswith('ocr_fallback.py'):
                    get_base_file(target)
                    
                    if name == 'write_to_file':
                        code = args.get('CodeContent') or args.get('codeContent')
                        while isinstance(code, str) and code.startswith('"'):
                            try: code = json.loads(code)
                            except: break
                        files[target] = code
                        
                    elif name == 'replace_file_content':
                        code = args.get('ReplacementContent') or args.get('replacementContent')
                        while isinstance(code, str) and code.startswith('"'):
                            try: code = json.loads(code)
                            except: break
                        lines = files[target].split('\n')
                        
                        start_str = args.get('StartLine') or args.get('startLine')
                        while isinstance(start_str, str) and start_str.startswith('"'):
                            try: start_str = json.loads(start_str)
                            except: break
                        start = int(start_str) - 1
                        
                        end_str = args.get('EndLine') or args.get('endLine')
                        while isinstance(end_str, str) and end_str.startswith('"'):
                            try: end_str = json.loads(end_str)
                            except: break
                        end = int(end_str)
                        
                        lines[start:end] = code.split('\n')
                        files[target] = '\n'.join(lines)
                        
                    elif name == 'multi_replace_file_content':
                        chunks = args.get('ReplacementChunks') or args.get('replacementChunks')
                        while isinstance(chunks, str):
                            try: chunks = json.loads(chunks)
                            except: break
                        
                        if not isinstance(chunks, list):
                            continue
                            
                        # Fix for chunks sometimes having string values with quotes
                        for chunk in chunks:
                            for key in ['StartLine', 'startLine', 'EndLine', 'endLine', 'ReplacementContent', 'replacementContent']:
                                if key in chunk:
                                    while isinstance(chunk[key], str) and chunk[key].startswith('"'):
                                        try: chunk[key] = json.loads(chunk[key])
                                        except: break
                                    
                        chunks.sort(key=lambda x: int(x.get('StartLine') or x.get('startLine')), reverse=True)
                        lines = files[target].split('\n')
                        for chunk in chunks:
                            code = chunk.get('ReplacementContent') or chunk.get('replacementContent')
                            start = int(chunk.get('StartLine') or chunk.get('startLine')) - 1
                            end = int(chunk.get('EndLine') or chunk.get('endLine'))
                            lines[start:end] = code.split('\n')
                        files[target] = '\n'.join(lines)

for target, content in files.items():
    print(f"Restoring {target}")
    with open(target, 'w', encoding='utf-8') as f:
        f.write(content)

print("Done restoring files.")
